presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java

```diff
@@ -197,7 +215,24 @@ public OptionalInt tryConvertToDirect(int maxDirectBytes)
         for (DictionaryRowGroup rowGroup : rowGroups) {
             directWriter.beginRowGroup();
             // todo we should be able to pass the stats down to avoid recalculating min and max
-            boolean success = tryConvertRowGroupToDirect(rowGroup.getValueCount(), rowGroup.getDictionaryIndexes(), maxDirectBytes);
+
+            boolean success;
+            int[] integerIndexes = rowGroup.getIntegerIndexes();
+            if (integerIndexes != null) {
+                success = tryConvertRowGroupToDirect(rowGroup.getValueCount(), integerIndexes, maxDirectBytes);
+            }
+            else {
+                short[] shortIndexes = rowGroup.getShortIndexes();
+                if (shortIndexes != null) {
+                    success = tryConvertRowGroupToDirect(rowGroup.getValueCount(), shortIndexes, maxDirectBytes);
+                }
+                else {
+                    byte[] byteIndexes = rowGroup.getByteIndexes();
+                    checkState(byteIndexes != null, "All 3 indexes are empty in dictionary");
+                    success = tryConvertRowGroupToDirect(rowGroup.getValueCount(), byteIndexes, maxDirectBytes);
+                }
+            }
```

Comment:
There are too many duplicated code in commit "Reduce dictionary row group size". Instead of having both the caller and the callee to be aware of the type + tunneling different types through the interface, can we have abstract class/interface over XXIndexes so that we only pass the interface?

Reply:
This code has lot of duplication and it has a lot to be desired.

First some background, there is a base class DictionaryColumnWriter and two dervied classes longDictionaryWriter and stringDictioanrywriter. (let us call them long and string writer).

Note: Java template code only works for Objects and does not work for primitives. (Unlike C++, which is code generation)

Let us consider tryConvertRowGroupToDirect method. Today this method is defined as
tryConvertRowGroupToDirect(int count, int/byte/short [] array, int maxBytes)

There are two implementations of this method in Long and String. So there could be virtual dispatch and since it is passing in an array the cost is amortized and negligible. Also there are some conditions that are done once per array and vectorization benefits.

We could define one method int tryConvertRowGroupToDirect(int index); This way the individual implementations of long and string need not worry about segments, but the virtual dispatch will impact the performance. 

So this is what I have settled on.

The DictionaryColumnWriter is responsible for null check of the segments and if there are multiple arrays in the segments it will iterate and call per individual array. But the long and string writer, have the same code copy/pasted 3 times just changing the parameter of the method.